core/env.go

```diff
@@ -0,0 +1,739 @@
+			rerr = fmt.Errorf("panic in GetSubchecks: %v %s", err, string(debug.Stack()))
+		}
+	}()
+
+	if len(check.Subchecks) > 0 {
+		return check.Subchecks, nil
+	}
+
+	if check.EnvironmentID != "" {
+		env, err := check.EnvironmentID.Decode()
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode environment for check %q: %w", check.Name, err)
+		}
+		resource, _, err := execEntrypoint(ctx, bk, progSock, pipeline, envCache, env, check.Name, nil)
+		if err != nil {
+			return nil, fmt.Errorf("failed to exec check environment container: %w", err)
+		}
+
+		if resource == nil {
+			// not a recursive check, there are no subchecks
+			return nil, nil
+		}
+		recursiveCheck, ok := resource.(*Check)
+		if !ok {
+			// not a recursive check, there are no subchecks
+			return nil, nil
+		}
+		return recursiveCheck.GetSubchecks(ctx, bk, progSock, pipeline, envCache)
+	}
+
+	return nil, nil
+}
+
+func (check *Check) Result(
+	ctx context.Context,
+	bk *buildkit.Client,
+	progSock string,
+	pipeline pipeline.Path,
+	envCache *EnvironmentCache,
+) (*CheckResult, error) {
+	if len(check.Subchecks) > 0 {
+		// This is a composite check, evaluate it by evaluating each subcheck
+		var eg errgroup.Group
+		success := true
+		var output string
+		for _, subcheck := range check.Subchecks {
+			subcheck := subcheck
+			eg.Go(func() error {
+				subresult, err := subcheck.Result(ctx, bk, progSock, pipeline, envCache)
+				if err != nil {
+					return fmt.Errorf("failed to get subcheck result for %q: %w", subcheck.Name, err)
+				}
+				if !subresult.Success {
+					success = false
+					output += fmt.Sprintf("Subcheck %q failed:\n%s\n", subcheck.Name, subresult.Output)
+				} else {
+					output += fmt.Sprintf("Subcheck %q succeeded:\n%s\n", subcheck.Name, subresult.Output)
+				}
+				return nil
+			})
+		}
+		if err := eg.Wait(); err != nil {
+			return nil, err
+		}
+		return &CheckResult{
+			Success: success,
+			Output:  "",
+		}, nil
+	}
+
+	if check.UserContainer != nil {
+		// check will be evaluated by exec'ing this container, with success based on exit code
+		ctr := check.UserContainer
+		success := true
+		var output string
+		_, evalErr := ctr.Evaluate(ctx, bk)
+		var execErr *buildkit.ExecError
+		switch {
+		case errors.As(evalErr, &execErr):
+			success = false
+			// TODO: really need combined stdout/stderr now
+			output = strings.Join([]string{evalErr.Error(), execErr.Stdout, execErr.Stderr}, "\n\n")
+		case evalErr != nil:
+			return nil, fmt.Errorf("failed to exec check user container: %w", evalErr)
+		default:
+			stdout, err := ctr.MetaFileContents(ctx, bk, progSock, "stdout")
+			if err != nil {
+				return nil, fmt.Errorf("failed to get stdout from check user container: %w", err)
+			}
+			stderr, err := ctr.MetaFileContents(ctx, bk, progSock, "stderr")
+			if err != nil {
+				return nil, fmt.Errorf("failed to get stderr from check user container: %w", err)
+			}
+			// TODO: really need combined stdout/stderr now
```

Comment:
Yep here's my thinking on this whole topic (broadly aligned with yours, and also a set of scattered loose thoughts 🙂):
1. It's important to our API that stdout/stderr are stored in the exec meta mount because that ties them to the cache of the exec.
   * Among other things, this ensures that our API works seamlessly with remote caching. E.g. `.Stdout()` is cached in the same way the filesystem of the exec is. 
   * This doesn't mean that we can't support streaming or anything. It's just an important implementation constraint to keep in mind for all of this.
2. We do indeed have the plumbing for streaming now, e.g. this method added as part of the previous re-arch mounts a file from buildkit's cache and streams it to the client (but for filesync): https://github.com/sipsma/dagger/blob/179525ff35a1e4c5b338472e47d350d4dbee8fb2/engine/buildkit/filesync.go#L254-L254
3. Whether we stream or return giant strings or both, there's either way some low level work needed to make combined stdout/stderr work as expected
   * Namely, in order to support all 3 APIs of `Stdout`, `Stderr` and `CombinedOutput`, we need to store the two streams of data in such a way that they can be presented both separately and together, and when combined it would preferably be interleaved as close as possible to the same order they were written to the pipes.
   * My thought has been that we might need to continue to use separate pipes for stdout/stderr, but then as we read lines from them, store the lines structured in some way that indicates which stream they were from (e.g. with some "header" like `[stdout]`/`[stderr]`) and then write them all to the same file with that structure.
   * That way, if you want just the stdout, you read the file, grab the stdout lines, strip the structure and return that stream. Same for stderr. And for combined, just read all the lines.
   * Probably other approaches too though.
4. I think you're spot on about the fact that being limited to "stdout/stderr of the most recent exec" being very limited.
   * While this would be tricky in the details, I can definitely see it making sense to be able to walk the DAG underlying a container and grab all the stdout/stderr/combined streams of the focused execs only. It's just a matter of figuring out stuff like whether we only grab the "most recent series of focused execs", "all focused execs", etc.
   * I also wouldn't want this to be the only API; it should hopefully just supplement the ability to get this data from an individual exec.

If you're in agreement about `3`, I think that we could at least implement that part as the foundation for everything else (whenever there's time obviously, it isn't exactly an emergency).